server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java

```diff
@@ -108,7 +110,15 @@ private void handleRequestBody(Buffer requestBody) {
         context.setRequestBody(requestBody);
 
         if (context.getResponseBody() == null) {
-            sendRequest();
+            proxy.getRateLimiter().limit(context, context.getRoute())
```

Comment:
You missed a scenario when future result of rateLimitResponse is failed.

Reply:
added `onFailure` handler